File: Top 150/lc_104.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self._recur_count(0,root)
    
    def _recur_count(self,count:int,node:Optional[TreeNode]):
        if node is None:
            return count 
        left = node.left
        right = node.right
        count += 1
        count1 = self._recur_count(count, left)
        count2 = self._recur_count(count, right)
        return max(count1,count2)


        

File: Top 150/lc_104.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
let maxDepth = function (root) {
  if (root === null) {
    return 0;
  }
  return _recurCount(0, root);
};

const _recurCount = (count, root) => {
  if (root === null) {
    return count;
  }
  let count1 = _recurCount((count += 1), root.left);
  let count2 = _recurCount((count += 1), root.right);
  return max(count1, count2);
};

class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

// Example test tree
let node7 = new TreeNode(7);
let node15 = new TreeNode(15);
let node20 = new TreeNode(20, node15, node7);
let node9 = new TreeNode(9);
let node3 = new TreeNode(3, node9, node20);

console.log(maxDepth(node3)); // expect 3
console.log(maxDepth(null)); // expect 0
console.log(maxDepth(new TreeNode(1))); // expect 1
